### PR TITLE
modifs on equational theories

### PR DESCRIPTION
**Pull request by ybertot**
According to discussions with @amahboubi, examples based on subtraction will be removed from this chapter.  To be done later.

### CODE REVIEW

File: tex/chProgramming.tex [hunk 1]
```diff
@@ -594,22 +594,24 @@ \section{Data types, first examples}
 \[ 0 + x \times (S\ 0) \]
```
Comment (darijgr) [inline tex/chProgramming.tex]: Do we need an empty line before this?

File: tex/chProgramming.tex [hunk 2]
```diff
@@ -594,22 +594,24 @@ \section{Data types, first examples}
...
 $S$ for natural numbers. More precisely, any variable-free boolean expression
 is equal to either $\top$ or $\bot$ modulo these axioms, and any
 variable-free expression in Peano arithmetic is equal either to $0$ or
-to an iterated application of the symbol $S$ to $0$. In both
+to an iterated application of the symbol $S$ to $0$.  In both
 cases, the other symbols in the signature represent functions, whose
-computational content is prescribed by the axioms.
+computational content is prescribed by the axioms.  In what follows,
+we shall call the distinctive symbols \emph{constructors}.  So the constructors
+in boolean arithmetic are \(\top\) and \(\bot\) and the constructors of
```
Comment (darijgr) [inline tex/chProgramming.tex]: Suggesting "in boolean arithmetic" -> "of boolean arithmetic" (see the end of this line).

File: tex/chProgramming.tex [hunk 3]
```diff
@@ -594,22 +594,24 @@ \section{Data types, first examples}
...
+Peano arithmetic are \(0\) and \(S\).
 
-In \Coq{}, such expressions are represented using a \emph{data type},
+In \Coq{}, such languages are represented using a \emph{data type},
 whose definition provides in a single declaration the name of the
-type, the symbols of (constants and) operations that \emph{build}
-elements of this type, plus some rules on how to compute on these
-elements. These data types are introduced by the means of an
+type and the constructors, plus some rules on how to compute on
+these constructors.  These data types are introduced by the means of an
```
Comment (darijgr) [inline tex/chProgramming.tex]: As commented elsewhere, I believe that "how to compute on these constructors" is unclear, and should rather be something like "how to define maps on these data types or prove theorems about their elements".

File: tex/chProgramming.tex [hunk 4]
```diff
@@ -962,6 +973,25 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
 what the \C{(addn n m)} program does is simply to pile \C{n} times the
 successor symbol on top of \C{m}.
 
+If we reflect again on the discussion of Peano arithmetics, as in
```
Comment (darijgr) [inline tex/chProgramming.tex]: "arithmetic" (singular)?

File: tex/chProgramming.tex [hunk 5]
```diff
@@ -962,6 +973,25 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+Section~\ref{sec:data}, we see that addition is provided by the definition of
+\C{addn}, and the usual axioms of Peano arithmetic, namely:
+\[S~ x + y = S (x + y) \qquad x + 0 = x \]
```
Comment (darijgr) [inline tex/chProgramming.tex]: This looks strange. Why are you destructing the first argument in the first equality and the second argument in the second? Shouldn't it either be (S x) + y = S (x + y) and 0 + y = y, or x + (S y) = S (x + y) and x + 0 = x ? (I'd also put a comma between the two equations.)

File: tex/chProgramming.tex [hunk 6]
```diff
@@ -962,6 +973,25 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+are actually provided by the computation behavior of the function,
+namely by the ``then'' branch and the ``else'' branch respectively.
+Therefore, Peano arithmetic is really provided by \Coq{} in two steps,
+first by providing the type of natural numbers and its constructors
+thanks to an \emph{inductive type definition}, and then by providing
```
Comment (darijgr) [inline tex/chProgramming.tex]: I'd replace "thanks to" by "using".

File: tex/chProgramming.tex [hunk 7]
```diff
@@ -962,6 +973,25 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+the operations as defined functions (usually recursive functions as in
+the case of addition).  The axioms are not constructed explicitely,
```
Comment (darijgr) [inline tex/chProgramming.tex]: "explicitly". This is a typo I've made many times myself :)

File: tex/chProgramming.tex [hunk 8]
```diff
@@ -962,6 +973,25 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+but they appear as part of the behavior of the addition function.  In
+practice, it will be able to create theorems whose statements are
```
Comment (darijgr) [inline tex/chProgramming.tex]: What is "it"?

**Comment by darijgr** [comment]
Nice changes. I've got my own partially conflicting changes in #30 . Instead of replacing the subn example by another one, I explicitly mention that the max(m-n, 0) semantics of subn is being used to define leq; IMHO this doesn't make it any worse an example.

File: tex/chProgramming.tex (hunk 5 above)
Comment (ybertot) [inline tex/chProgramming.tex]: You are right, this is a mistake

**Comment by gares** [comment]
> Instead of replacing the subn example by another one

We will address this issue in another PR

**Pull request merged**